bin/util/common.sh

```diff
@@ -1,36 +1,60 @@
+# a file to write captured warnings to
+# it cannot be a variable, because the warnings function may be used in a pipeline, which causes a subshell, which can't modify parent scope variables
+_captured_warnings_file=$(mktemp -t heroku-buildpack-php-captured-warnings-XXXX)
+
 error() {
+	# send all of our output to stderr
+	exec 1>&2
 	# if arguments are given, redirect them to stdin
 	# this allows the funtion to be invoked with a string argument, or with stdin, e.g. via <<-EOF
 	(( $# )) && exec <<< "$@"
-	echo
+	echo -e "\033[1;31m" # bold; red
 	echo -n " !     ERROR: "
 	# this will be fed from stdin
-	indent no_first_line_indent
-	echo
+	indent no_first_line_indent " !     "
+	if [[ -s "$_captured_warnings_file" ]]; then
+		echo "" | indent "" " !     "
+		echo "REMINDER: the following warnings were emitted during the build;" | indent "" " !     "
+		echo "check the details above, as they may be related to the error:" | indent "" " !     "
+		cat "$_captured_warnings_file" | indent "" " !     - "
+	fi
+	echo -e "\033[0m" # reset style
 	exit 1
 }
 
 warning() {
+	# send all of our output to stderr
+	exec 1>&2
 	# if arguments are given, redirect them to stdin
 	# this allows the funtion to be invoked with a string argument, or with stdin, e.g. via <<-EOF
 	(( $# )) && exec <<< "$@"
-	echo
+	echo -e "\033[1;33m" # bold; yellow
```

Comment:
I do that with notices. Hmmh. I'd like to try this for now I think; the warnings are very selectively "this is a really bad practive because it can get you into trouble", whereas notices are like "it would be nice if you..."

Reply:
Ahh, you've got extra levels of warnings. If this is fairly severe, then I think it's fine.